janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -414,63 +423,50 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
+                        .setPageSize(this.storeManager.getPageSize())
+                        .setConsistencyLevel(getTransaction(txh).getReadConsistencyLevel()).build())))
                 .getOrElseThrow(EXCEPTION_MAPPER);
     }
 
     /**
      * This class provides a paging implementation that sits on top of the DSE Cassandra driver. The driver already
      * has its own built in paging support but this has limitations when doing a full scan of the key ring due
      * to how driver paging metadata is stored. The driver stores a full history of a given query's paging metadata
-     * which can lead to OOM issues on non-trivially sized data sets. This class overcomes this by doing another level
-     * of paging that re-executes the query after a configurable number of rows. When the original query is re-executed
-     * it is initialized to the correct offset using the last page's metadata.
+     * which can lead to OOM issues on non-trivially sized data sets. This class overcomes this by forcing the internal
+     * metadata which isn't needed anymore to be deleted to be free for GC.
      */
     private class CQLPagingIterator implements Iterator<Row> {
 
-        private ResultSet currentResultSet;
-
-        private int index;
-        private int paginatedResultSize;
-        private final Supplier<Statement> statementSupplier;
-
-        private byte[] lastPagingState = null;
+        private final List<ExecutionInfo> dseStoredExecutionInfos;
+        private final java.util.Iterator<Row> currentPageIterator;
 
-        public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSupplier) {
-            this.index = 0;
-            this.paginatedResultSize = pageSize;
-            this.statementSupplier = statementSupplier;
-            this.currentResultSet = getResultSet();
+        public CQLPagingIterator(BoundStatement boundStatement) {
+            ResultSet currentResultSet = session.execute(boundStatement);
+            currentPageIterator = currentResultSet.iterator();
+            this.dseStoredExecutionInfos = currentResultSet.getExecutionInfos();
         }
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            return currentPageIterator.hasNext();
         }
 
         @Override
         public Row next() {
-            if(index == paginatedResultSize) {
-                currentResultSet = getResultSet();
-                this.index = 0;
-            }
-            this.index++;
-            lastPagingState = currentResultSet.getExecutionInfo().getPagingStateUnsafe();
-            return currentResultSet.one();
-
-        }
-
-        private ResultSet getResultSet() {
-            final Statement boundStmnt = statementSupplier.get();
-            if (lastPagingState != null) {
-                boundStmnt.setPagingStateUnsafe(lastPagingState);
+            if(dseStoredExecutionInfos.size()>1){
```

Comment:
Current DSE `ResultSet` implementations return mutable `ArrayList` when `getExecutionInfos()` is called. The only case when this list is Immutable is when `SinglePageResultSet` is used but we won't mutate that list because it always has the size of `1`. Thus, it's not possible to enter this `if` statement with that Immutable list.